### PR TITLE
Alternative name component used for faster comparisons

### PR DESCRIPTION
**Pull request by lassade**
`Name` is a better alternative to name entities instead of using `Option<String>`, because it allows for faster searches. I'm currently using this component for find entities in deep nested hierarchies in my animation system;

It works by also storing a hash along side name, then when peforming a comparison between two `Name`s it first check if the hashes matches, if don't the strings are different and the acctual string comparision is skipped.

I'm using ahash because it's the fastest hasing algo that I know of;

### CODE REVIEW

**Comment by smokku** [comment]
A candidate for `bevy_extras` crate?

**Comment by lassade** [comment]
This could be the default way of naming entites, thats why I put it on bevy_core

File: crates/bevy_core/Cargo.toml [hunk 1]
```diff
@@ -21,3 +21,6 @@ bevy_math = { path = "../bevy_math", version = "0.4.0" }
 bevy_reflect = { path = "../bevy_reflect", version = "0.4.0", features = ["bevy"] }
 bevy_tasks = { path = "../bevy_tasks", version = "0.4.0" }
 bevy_utils = { path = "../bevy_utils", version = "0.4.0" }
+
+# other
+ahash = "0.6.2"
```
Comment (memoryruins) [inline crates/bevy_core/Cargo.toml:26]: `bevy_utils` re-exports `ahash::AHasher` as `bevy_utils::AHasher`, so we can use it from there instead.

**Review by memoryruins — CHANGES_REQUESTED** [review]

File: crates/bevy_core/Cargo.toml (hunk 1 above)
Comment (lassade) [reply crates/bevy_core/Cargo.toml]: Oh nice!

File: crates/bevy_core/src/name.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    hash::{Hash, Hasher},
+    ops::Deref,
+};
+
+/// Component used to identify a entity. Stores a hash for faster comparisons
+#[derive(Debug, Clone, Reflect)]
+#[reflect(Component)]
+pub struct Name {
+    hash: u64, // TODO: Shouldn't be serialized
+    name: String,
```
Comment (DJMcNab) [inline crates/bevy_core/src/name.rs:13]: Could this be `std::borrow::Cow<'static, str>`, or could this whole struct be more general then use:
```rust
type Name = FastCompare<String>;
```

**Comment by cart** [comment]
Can you think of a reason not to just extend the existing `Labels` system to cache hashes for faster comparisons?

**Comment by lassade** [comment]
No other that a name must be one per entity; We could so something similar to `Labels` but I'm not sure how that will look like;

File: crates/bevy_core/src/name.rs (hunk 2 above)
Comment (lassade) [reply crates/bevy_core/src/name.rs:13]: Yeah it could

File: crates/bevy_core/src/name.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+impl From<&str> for Name {
+    #[inline(always)]
+    fn from(name: &str) -> Self {
+        Name::new(name.to_owned())
+    }
+}
+
+impl Hash for Name {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
```
Comment (lassade) [inline crates/bevy_core/src/name.rs:61]: It will be correct to use `self.hash.hash(state)`?

**Comment by cart** [comment]
I do think this is probably a good candidate for a bevy_extra crate. I'm not yet convinced we need "canonical names" built in to the engine by default. "selecting by string name" isn't a general pattern we should be encouraging (we should encourage marker components), and the cases where it makes sense are very context specific (and therefore should probably live in game code). Even the Labels system we currently have should probably live in bevy_extra (at least for now). I added them for some debugging scenarios and because Godot has a similar feature, but i think the Labels system will only really start getting useful in the context of a visual scene editor. 

I think its time to get bevy_extra rolling. It would be a separate repo (to draw hard lines between the projects and denote a different level of support).

**Comment by mockersf** [comment]
to add context on why `Name` (or `Labels`) can be useful as a fully supported component, as I had the "need" for them:
* when loading a gltf file, names can be specified from the file, and a marker component can't be generated for them. this is related to the scene need you mention
* @lassade is working on animations, where one can specify the target for an animation by a path made of names

**Comment by cart** [comment]
Yeah thats fair. And I guess canonical names are good for visual scene organization (ex the identifier used when rendering entity hierarchies in the editor).

In general I have a reflex to not add something like this until we have a concrete scenario that uses it. But I guess it probably makes sense for GLTF-loaded entities to use the GLTF node name as the "canonical name displayed in the bevy editor hierarchy viewer".

Ok I think im convinced that we should have this in-engine.

File: crates/bevy_core/src/name.rs (hunk 3 above)
Comment (cart) [reply crates/bevy_core/src/name.rs:61]: I think we could probably get away with that. Its effectively double-hashing, so our collision risk almost certainly goes up.

But short of implementing a custom hasher for every Name use case, im not sure we have a better option.

File: crates/bevy_core/src/name.rs (hunk 2 above)
Comment (cart) [reply crates/bevy_core/src/name.rs:13]: Im cool with making this less general for now to cut down on the amount of abstraction. If we find multiple use cases later and it seems worth it, we can always factor this out.

File: crates/bevy_core/src/name.rs (hunk 3 above)
Comment (cart) [reply crates/bevy_core/src/name.rs:61]: Eh actually I'd rather be safe now and optimize later if we prove its not too risky.

**Pull request merged**